packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -653,7 +655,24 @@ export function getValidatorApi({chain, config, logger, metrics, network, sync}:
     },
 
     async registerValidator(registrations) {
-      return chain.executionBuilder?.registerValidator(registrations);
+      // should only send active or pending validator to builder
+      // Spec: https://ethereum.github.io/builder-specs/#/Builder/registerValidator
+      const headState = chain.getHeadState();
+      const currentEpoch = chain.clock.currentEpoch;
+      const seenValidators = new Set<ValidatorIndex>();
+
+      const filteredRegistrations = registrations.filter((registration) => {
+        const {pubkey} = registration.message;
+        const validatorIndex = headState.epochCtx.pubkey2index.get(pubkey);
+        if (validatorIndex === undefined || seenValidators.has(validatorIndex)) return false;
```

Comment:
seems like `seenValidators` is only deduping the request from what it seems and the purpose of `isActiveValidator`/`isPendingValidator` is to filter out exited/slashed validators?

Reply:
i don't think we need `seenValidators` processing as this particular requests would/should not contain any duplicates

Also should the `ValidatorStatus` be generated/maintained as part of CachedBeaconState generation? that would be optimized to generate and keep it as well as serve it to the validator client, which can also track/logout the groupings of the validators keys w.r.t. statues

Reply:
I dropped `seenValidators`. Since this really goes with `ValidatorStatus` in the api spec, I use `getValidatorStatus()` util to avoid adding another function to check for pending validator.

Regarding adding `ValidatorStatus` to `CachedBeaconState`, it'd introduce some memory increase because we need to add that for all validators, I don't see the need for now